CLDR-15034 Keyboard v3.0 draft DTD (#1847)
unicode-org/cldr: docs/ldml/tr35-keyboards.md

The CLDR keyboard format seeks to address these challenges, by providing an interchange format for the communication of keyboard mapping data independent of vendors and platforms. Keyboard authors can then create a single mapping file for their language, which implementations can use to provide that language’s keyboard mapping on their own platform.

Additionally, the standardized identifier for keyboards can be used to communicate, internally or externally, a request for a particular keyboard mapping that is to be used to transform either text or keystrokes. The corresponding data can then be used to perform the requested actions.  For example, a remote screen-access application (such as used for customer service or server management) would be able to communicate and choose the same keyboard layout on the remote device as is used in front of the user, even if the two systems used different platforms.

The data can also be used in analysis of the capabilities of different keyboards. It also allows better interoperability by making it easier for keyboard designers to see which characters are generally supported on keyboards for given languages.

> To illustrate this specification, here is an abridged layout representing the English US 101 keyboard on the macOS operating system (with an inserted long-press example).

Author: I'm going to remove the example, again, for now.

3. Allow platforms to be able to use CLDR keyboard data for the character-emitting keys (non-frame) aspects of keyboard layouts.
    * For example, platform-specific keys such as Fn, Numpad, IME swap keys, and cursor keys are out of scope.
    * This also means that modifier (frame) keys cannot generate output, such as capslock -> backslash.
4. Deprecate & archive existing LDML platform-specific layouts so they are not part of future releases

<!--
1. Make the XML as readable as possible.
2. Represent faithfully keyboard data from major platforms: it should be possible to create a functionally-equivalent data file (such that given any input, it can produce the same output).
3. Make as much commonality in the data across platforms as possible to make comparison easy. -->

Some non-goals (outside the scope of the format) currently are:

1. Adaptation for screen scaling resolution. Instead, keyboards should define layouts based on physical size. Platforms may interpret physical size definitions and adapt for different physical screen sizes with different resolutions.
2. Unify the world's vkey and scan code mapping information

> 3. Unify pre-existing platform layouts themselves (i.e. existing fr-azerty on platform a, b, c)

Author: n.b.!

**Key:** A key on a physical keyboard, or a virtual key on an on-screen layout.

**Key code:** The integer code sent to the application on pressing a key.

**Key map:** The basic mapping between hardware or on-screen positions and the output characters for each set of modifier combinations associated with a particular layout. There may be multiple key maps for each layout.

**Keyboard:** A particular arrangement of keys for the inputting of text, such as either a physical or virtual keyboard.

**Keyboard author:** The person or group of people designing and producing a particular keyboard layout designed to support one or more languages. In the context of this specification, that author may be editing the LDML XML file directly or by means of software tools.

**Keyboard layout:** A layout is the overall keyboard configuration for a particular locale. Within a keyboard layout, there is a single base map, one or more key maps and zero or more transforms.

> **Layer** is an arrangement of keys on a virtual keyboard. Since it is often not intended to use two hands on a visual keyboard to allow the pressing of modifier keys. Modifier keys are made sticky in that one presses one, the visual representation, and even arrangement, of the keys change, and you press the key. This visual representation is a layer. Thus a virtual keyboard is made up of a set of layers.

Author: That is simpler, thanks

2. Some features, such as multitap and flicks, have the potential to reduce accessibility and thus should be discouraged. For example, multitap requires pressing keys at a certain speed, and flicks require a more complex movement (press-and-flick) beyond a simple tap.
3. Public feedback is welcome on any aspects of this document which might hinder accessibility.

## 3 <a name="Definitions" href="#Definitions">Definitions</a>

**Arrangement** is the term used to describe the relative position of the rectangles that represent keys, either physically or virtually. A physical keyboard has a static arrangement while a virtual keyboard may have a dynamic arrangement that changes per language and/or layer. While the arrangement of keys on a keyboard may be fixed, the mapping of those keys may vary.

**Base character:** The character emitted by a particular key when no modifiers are active. In ISO terms, this is group 1, level 1.

**Base map:** A mapping from the positions to the base characters. There is only one base map per layout. The characters on this map can be output by not using any modifier keys.

> **Core keys:** also known as “alpha” block. The primary set of key values on a keyboard that are used for typing the target language of the keyboard. For example, the three rows of letters on a standard US QWERTY keyboard (QWERTYUIOP, ASDFGHJKL, ZXCVBNM) together with the most significant punctuation keys. Usually this equates to the minimal keyset for a language as seen on mobile phone keyboards.

Reviewer: While I realize that we don't appear to use the term elsewhere within the document yet, lines 144-146 reference the term "frame keys" and imply a definition.

We could make that more explicit here, and the contrast with that definition would help further clarify the "core key" definition seen here.

Author: Good idea. 

- [ ] Add frame keys to definitions.

Author: I added frame keys.

<!-- The CLDR keyboard format provides for the communication of keyboard mapping data between different modules, and the comparison of data across different vendors and platforms. The standardized identifier for keyboards can be used to communicate, internally or externally, a request for a particular keyboard mapping that is to be used to transform either text or keystrokes. The corresponding data can then be used to perform the requested actions.

For example, a web-based virtual keyboard may transform text in the following way. Suppose the user types a key that produces a "W" on a QWERTY keyboard. A web-based tool using an AZERTY virtual keyboard can map that text ("W") to the text that would have resulted from typing a key on an AZERTY keyboard, by transforming "W" to "Z". Such transforms are in fact performed in existing web applications.

The data can also be used in analysis of the capabilities of different keyboards. It also allows better interoperability by making it easier for keyboard designers to see which characters are generally supported on keyboards for given languages.

> To illustrate this specification, here is an abridged layout representing the English US 101 keyboard on the Mac OS X operating system (with an inserted long-press example). For more complete examples, and information collected about keyboards, see keyboard data in XML.

Reviewer 2: ```suggestion
To illustrate this specification, here is an abridged layout representing the English US 101 keyboard on the macOS operating system (with an inserted long-press example). For more complete examples, and information collected about keyboards, see keyboard data in XML.
```

They keep changing the name of the operating system. I believe this is current, per Apple's website: https://www.apple.com/macos/monterey/

>
> Occurrence: required, single
>
> </small>

_Attribute:_ `conformsTo` (required)

This attribute distinguishes the keyboard from prior versions,
and it also specifies the minimum CLDR version required.

For purposes of this current draft spec, the value should always be `techpreview`

```xml
<keyboard … conformsTo="techpreview"/>
```

_Attribute:_ `locale` (required)

This mandatory attribute represents the primary locale of the keyboard using Unicode locale identifiers (see [LDML](tr35.md)) - for example `"el"` for Greek. Sometimes, the locale may not specify the base language. For example, a Devanagari keyboard for many languages could be specified by BCP-47 code: `"mul-Deva"`. For further details, see [Keyboard IDs](#Keyboard_IDs).

**Example** (for illustrative purposes only, not indicative of the real data)

> 

Author: Good point. I'll change these to `ka` and `fr-CH-t-k0-azerty` respectively.

> 

Author: done

  …
</keyboard>
```

```xml

> <keyboard locale="fr-CH-t-k0-android">

Reviewer: Given the work we've done on the spec, and our goal to facilitate cross-platform keyboards, I find it concerning that our two examples for locale both include a platform name, which would imply non-cross-platform design.

_One_ is fine to show support for older keyboard designs that weren't made with this in mind, but having both of our only two examples here do this is a bit concerning due to the inferences an uninitiated reader may draw.

Author: - [ ] Make non platform ids more prominent. 

Yes. This is the balance between leveraging what was written before and starting over. I thought about that while linking in this section. We should prefer non platform ids.

The optional `<locales>` element allows specifying additional or alternate locales. Denotes intentional support for an extra language, not just that a keyboard incidentally supports a language’s orthography.

**Syntax**

```xml
<locales>
    <locale id="…"/>
    <locale id="…"/>
</locales>
```

> <small>
>
> Parents: `keyboard`
> Children: `locale`
> Occurrence: optional, single
>
> </small>

### 5.y <a name="Element_locale" href="#Element_locale">Element: locale</a>

The optional `<locales>` element allows specifying additional or alternate locales. Denotes intentional support for an extra language, not just that a keyboard incidentally supports a language’s orthography.

**Syntax**

```xml
<locale id="{id}"/>
```

> <small>
>
> Parents: `locales`
> Children: _none_
> Occurrence: optional, multiple
>
> </small>

_Attribute:_ `id` (required)

> The BCP47 locale ID of an additional language supported by this keyboard.
> Do _not_ include the `-k0-` subtag for this additional language.

**Example**

See [Principles for Keyboard IDs](#Principles_for_Keyboard_IDs) for discussion and further examples.

```xml
<!-- Pan Nigerian Keyboard-->

> <keyboard locale="mul-Latn-NG-t-k0-panng">

Reviewer: I realize that _this_ example doesn't use a platform name in its locale string, in reference to my prior comment - but it's far enough after the section that I don't believe it sufficient to address that comment's concerns.  It's not in direct visual line-of-sight for comparison + contrast against the two examples seen at the part most directly defining the property.

Author: I think  have de emphasized platform now.

>
> Parents: [keyboard](#Element_keyboard)
> Children: _none_
> Occurrence: required, single
>
> </small>

<!-- _Attribute:_ `platform` (required)

> > The platform source version. Specifies what version of the platform the data is from. For example, data from Mac OS X 10.4 would be specified as `platform="10.4"`. For platforms that have unstable version numbers which change frequently (like Linux), this field is set to an integer representing the iteration of the data starting with `"1"`. This number would only increase if there were any significant changes in the keyboard data. -->

Reviewer 2: Not changing this ref to Mac OS X because 10.4 was "Mac OS X Tiger"

_Attribute:_ `number` (required)

> Must be a [semver-compatible](https://semver.org) version number.

> <!-- TODO make the above into a reverence to SEMVER -->

Reviewer: Minor nit:  this TODO is addressed now, isn't it?  Or do we want to expound on this reference further?

Author: It's an editorial note. Unicode has a certain structure for making stable references to external standards. Just reminding myself so I don't forget.

_Attribute:_ `transformPartial="hide"` (optional)

> This attribute describes the behavior of the system while in a transform. When this attribute is present then don't show the values of the buffer as the user is typing a transform (this behavior can be seen on Windows or Linux platforms).

> By default (when this attribute is not present), show the values of the buffer as the user is typing a transform (this behavior can be seen on the Mac OS X platform).

Reviewer 2: ```suggestion
By default (when this attribute is not present), show the values of the buffer as the user is typing a transform (this behavior can be seen on the macOS platform).
```

If this attribute is present, it must have a value of omit.

_Attribute:_ `transformPartial="hide"`

> This attribute describes the behavior of the system while in a transform. When this attribute is present then don't show the values of the buffer as the user is typing a transform (this behavior can be seen on Windows or Linux platforms).

By default (when this attribute is not present), show the values of the buffer as the user is typing a transform (this behavior can be seen on the macOS platform).

> If this attribute is present, it must have a value of hide.

Author: There is the question about whether the platform should default to hide for these layouts.  I added this to the agenda.

> If this attribute is present, it must have a value of hide.

Author: I'm going to mark this resolved from the spec draft point of view and we can discuss next time we meet

* * *

### 5.8 <a name="Element_keys" href="#Element_keys">Element: keys</a>

This element defines the properties of all possible keys via [`<key>` elements](#Element_key) used in all layouts.
It is a “bag of keys” without specifying any ordering or relation between the keys.
There is only a single `<keys>` element in each layout.

**Syntax**

```xml
<keys>
    <key … />
    <key … />
    <key … />
    <flicks … />
    <key … />
    <flicks … />
</keys>
```

`key` and `flicks` elements may be interleaved in any order.

> <small>
>
> Parents: [keyboard](#Element_keyboard)
> Children: [key](#Element_key), [flicks](#Element_flicks)
> Occurrence: required, single
>
> </small>

* * *

### 5.9 <a name="Element_key" href="#Element_key">Element: key</a>

This element defines a mapping between an abstract key and its output. This element must have the `keys` element as its parent. The `key` element is referened by the `keys=` attribute of the [`row` element](#Element_row).

**Syntax**

```xml
<key
 id="{key id}"
 [flicks="{flicks identifier}"]

>  [gap="true"]

Reviewer: Please contrast against [en-t-k0-teletype.xml](https://github.com/unicode-org/cldr/pull/1847/files#diff-9b1c99aec1ff6f44a068a703a57c0a534183941f44db9128771fff610cc9d0e1) as seen in this PR.

Here, gap is given a boolean type, but in that file, it's numeric.

Author: I Didn't get to either the dtd or the other files yet.

 [longPress="{long press keys}"]
 [longPressDefault="{default longpress target}"]
 [multitap="{the output on subsequent taps}"]
 [switch="{layer id}"]
 [to="{the output}"]
 [transform="no"]
 [width="{key width}"]
 /><!-- {Comment to improve readability (if needed)} -->
```

> <small>
>
> Parents: [keys](#Element_keys)
> Children: _none_
> Occurrence: optional, multiple
> </small>

**Note**: The `id` attribute is required, as is _at least one of_ `switch`, `gap`, or `to`.

_Attribute:_ `id`

> The `id` attribute uniquely identifies the key. NMTOKEN, restricted to "[a-zA-Z0-9_.-]". It can (but needn't be) the Latin key name for a Latn script keyboard (a, b, c, A, B, C, …), or any other valid token (e-acute, alef, alif, alpha, …)

_Attribute:_ `flicks="flick-id"` (optional)

> The `flicks` attribute indicates that this key makes use of a [`flicks`](#Element_flicks) set with the specified id.

_Attribute:_ `gap="true"` (optional)

> > The `gap` attribute indicates that this key does not have any appearance, but represents a "gap" of the specified number of key widths. Can be used with `width` to set a width.

Reviewer: Same as prior comment:  contrast with [en-t-k0-teletype.xml](https://github.com/unicode-org/cldr/pull/1847/files#diff-9b1c99aec1ff6f44a068a703a57c0a534183941f44db9128771fff610cc9d0e1), ensure consistency.

Author: Gap=true is what we discussed in committee.

```xml
<key id="mediumgap" gap="true" width="1.5"/>
```

_Attribute:_ `longPress="optional"` (optional) (discouraged, see [Accessibility](#Accessibility))

> The `longPress` attribute contains any characters that can be emitted by "long-pressing" a key, this feature is prominent in mobile devices. The possible sequences of characters that can be emitted are whitespace delimited. Control characters, combining marks and whitespace (which is intended to be a long-press option) in this attribute are escaped using the `\u{...}` notation.

_Attribute:_ `longPressDefault` (optional)

> Indicates which of the `longPress` target characters is the default long presstarget, which could be different than the first element. Ignored if not in the `longPress` list. Characters in this attribute can be escaped using the `\u{...}` notation.
> For example, if the `longPressDefault` is a key whose [display](#Element_displayMap) appears as `{` an implementation might render the key as follows:
>
> ![keycap hint](images/keycapHint.png)

> _Attribute:_ `multitap` (optional) (discouraged, see [Accessibility]

Reviewer 2: ```suggestion
_Attribute:_ `multitap` (optional) (discouraged, see [Accessibility])
```

> Also note that `switch=` is ignored for hardware layouts: their shifting is controlled via
> the modifier keys.

_Attribute:_ `to` (required)

> The `to` attribute contains the output sequence of characters that is emitted when pressing this particular key. Control characters, whitespace (other than the regular space character) and combining marks in this attribute are escaped using the `\u{...}` notation. More than one key may output the same output.

_Attribute:_ `transform="no"` (optional)

> The `transform` attribute is used to define a key that never participates in a transform but its output shows up as part of a transform. This attribute is necessary because two different keys could output the same characters (with different keys or modifier combinations) but only one of them is intended to be a dead-key and participate in a transform. This attribute value must be no if it is present.
>
> For example, suppose there are the following keys, their output and one transform:

```xml
<keys>
    <key id="X" to="^" transform="no"/>
    <key id="OptX" to="^"/>
</keys>
…
<transforms …>
    <transform from="^e" to="ê"/>
</transforms>
```

> * **X** outputs `^` (the Carat)

Reviewer 2: ```suggestion
* **X** outputs `^` (caret)
```

* Option-**X** outputs `^` but is intended to be the first part of a transform.
* Option-**X** + `e` → `ê`

> Without the `transform="no"` on the base key **X**, it would not be possible to

> > type the sequence `^e` (carat+e) as it would turn into `ê` per the transform.

Reviewer 2: ```suggestion
> type the sequence `^e` (caret+e) as it would turn into `ê` per the transform.
```

> The `gap` attribute indicates that this key does not have any appearance, but represents a "gap" of the specified number of key widths. Can be used with `width` to set a width.

```xml
<key id="mediumgap" gap="true" width="1.5"/>
```

> _Attribute:_ `longPress="a b c"` (optional) (discouraged, see [Accessibility](#Accessibility))

Reviewer 2: ```suggestion
_Attribute:_ `longPress="a b c"` (optional) (discouraged, see [Accessibility])
```

Is the `(#...)` needed? You aren't using it elsewhere.

* `names`
* `settings`

-->

* * *

### <a name="Element_displayMap" href="#Element_displayMap">Element: displayMap</a>

The displayMap can be used to describe what is to be displayed on the keytops for various keys. For the most part, such explicit information is unnecessary since the `@to` element from the `keys/key` element can be used. But there are some characters, such as diacritics, that do not display well on their own and so explicit overrides for such characters can help.
Another useful scenario is where there are doubled diacritics, or multiple characters with spacing issues.

The `displayMap` consists of a list of display subelements.

displayMaps are designed to be shared across many different keyboard layout descriptions, and imported with `<import>` in where needed.

> For combining characters, U+25CC `◌` is used as a base.

Reviewer 2: ```suggestion
For combining characters, U+25CC `◌` is used as a base. It is an error to use a combining character without a base in the `display` attribute.
```

Author: Good suggestion. There may be an existing Unicode term for this.

> to choose a different keyboard file, or use some other fallback behavior when using a
> hardware keyboard.
>
> When using an on-screen keyboard, if there is not a `<layerMaps form="touch">`
> element, the `<layerMaps form="hardware">` element can be used for on-screen use.

### 5.13.1 <a name="Element_layerMap" href="#Element_layerMap">Element: layerMap</a>

A `layerMap` element describes the configuration of keys on a particular layer of a keyboard. It contains one or more `row` elements to describe which keys exist in each row.

**Syntax**

```xml
<layerMap id="layerId" modifier="{Set of Modifier Combinations}">
    ...
</layerMap>
```

> <small>
>
> Parents: [keyboard](#Element_keyboard)
> Children: [row](#Element_row)
> Occurrence: optional, multiple
>
> </small>

_Attribute_ `id` (required for `touch`)

> The `id` attribute identifies the layer for touch layouts.  This identifier specifies the layout as the target for layer switching, as specified by the `switch=` attribute on the [`<key>`](#Element_key) element.
> Touch layouts must have one `layerMap` with `id="base"` to serve as the base layer.

_Attribute:_ `modifier` (required for `hardware`)

> > This has two roles. It acts as an identifier for the `layer` element for hardware keyboards (in the absence of the id= element) and also provides the linkage from the hardware modifiers into the correct `layer`. To indicate that no modifiers apply the reserved name of "none" can be used. For the purposes of fallback vkey mapping, the following modifier components are reserved: "shift", "ctrl", "alt", "caps", "cmd", "opt" along with the "L" and "R" optional single suffixes for the first 3 in that list.

Reviewer: > To indicate that no modifiers apply the reserved name...

A comma would be appreciated between "modifiers" and "apply".  That sentence was a bit tricky to parse on first read.

Author: - [ ] Comma

    <display …/>
    <displayOptions baseCharacter="x"/>
</displayMap>
```

> <small>
>
> Parents: [displayMap](#Element_displayMap)
>
> Children: _none_
>
> Occurrence: optional, single
>
> </small>

_Attribute:_ `baseCharacter` (optional)

**Note:** At present, this is the only option settable in the `displayOptions`.

> Some scripts/languages may prefer a different base than U+25CC.
> For Lao for example, `x` is often used as a base instead of `◌`.
> Setting `baseCharacter="x"` (for example) is a _hint_ to the implementation which
> requests U+25CC to be substituted with `x` on display.
> As a hint, the implementation may ignore this option.
>
> **Note** that not all base characters will be suitable as bases for combining marks.

* * *

### <a name="Element_layerMaps" href="#Element_layerMaps">Element: layerMaps</a>

This element represents a set of `layerMap` elements with a specific physical form factor, whether
hardware or touch layout.

> <small>
>
> Parents: [keyboard](#Element_keyboard)
>
> Children: [import](#Element_import), [layerMap](#Element_layerMap), [_special_](tr35.md#special)
>
> Occurrence: required, multiple
>
> </small>

_Attribute:_ `form` (required)

> > `form` may be either `hardware` or `touch`.

Reviewer 2: I note that our glossary says "physical keyboard" and "virtual keyboard" but here we use "hardware" and "touch". Hmmm.

Author: - [ ] glossary should be harmonized.

Author: harmonized.

_Attribute:_ `to` (required)

> This attribute represents the characters that are output from the transform. The output can contain more than one character, so you could have `<transform from="´A" to="Fred"/>`.

Control characters, whitespace (other than the regular space character) and combining marks in this attribute are escaped using the `\u{...}` notation.

> 

Reviewer 3: Let's remove the platform names from these examples.

Author: good idea.

### 5.20 <a name="Element_reorder" href="#Element_reorder">Element: reorders, reorder</a>

The reorder transform is applied after all transforms except for those with `type="final"`.

> This transform has the job of reordering sequences of characters that have been typed, from their typed order to the desired output order. The primary concern in this transform is to sort combining marks into their correct relative order after a base, as described in this section. The reorder transforms can be quite complex, keyboard layouts will almost always import them.

Reviewer: > The reorder transforms can be quite complex, keyboard layouts will almost always import them.

Glad to see this sentence here.

That said, we may want to ensure it sticks out in some way if we expect users (i.e. keyboard authors) to reference this document with any regularity; with it being the last sentence in the paragraph, I could see people skimming past it easily.

If this is targeted more at developers of platforms that may look to implement support for LDML keyboards, it's probably fine as is.

Author: It's both, somewhat aimed at implementers first.  We have a number of comments about best practices, but definitely not a good on-ramp for authors. 

These don't strictly belong in the spec, but they could be adjacent documents:

- [ ] Start a draft guide for authors. 
- [ ] Start a draft guide for implementers.

Author: @jahorton  created tickets for the above here: -> https://unicode-org.atlassian.net/issues/?jql=issueKey%20in%20(CLDR-15847%2CCLDR-15848)

> 

Reviewer 3: I thought we agreed to remove the "after" attribute.

Author: true. will remove.

| xY? ⋁ x ⇒ xY?                              |                                              |
| xLY? ⋁ x ⇒ xLY?                            |                                              |
| xLY ⋁ x ⇒ xLY?                             |                                              |
-->

* * *

## <a name="Keyboard_IDs" href="#Keyboard_IDs">Keyboard IDs</a>

There is a set of subtags that help identify the keyboards. Each of these are used after the `"t-k0"` subtags to help identify the keyboards. The first tag appended is a mandatory platform tag followed by zero or more tags that help differentiate the keyboard from others with the same locale code.

### <a name="Principles_for_Keyboard_IDs" href="#Principles_for_Keyboard_IDs">Principles for Keyboard IDs</a>

The following are the design principles for the IDs.

1. BCP47 compliant.
   1. Eg, `en`, `sr-Cyrl`, or `en-t-k0-extended`.
2. Use the minimal language id based on `likelySubtag`s.

>    1. Eg, instead of `en-US`, use `en`, and instead of `fr-Latn-FR` use `fr`. Because there is `<likelySubtag from="en" to="en_Latn_US"/>`, en-US → en.

Reviewer 3: This example is confusing because the `likelySubtag` covers only the English example and straddles the French example.

Author: Will change to the following, and link to the likely subtags documentation.

> Eg, instead of `fa-Arab`, use `fa`.

   2. The data is in <https://github.com/unicode-org/cldr/tree/main/common/supplemental/likelySubtags.xml>

> 3. Keyboard files should be platform-independent, however, a platform id is the first subtag after `-t-k0-` if present. If a keyboard on the platform changes over time, both are dated, eg `bg-t-k0-chromeos-2011`. When selecting, if there is no date, it means the latest one.

Reviewer 3: Rephrase:
Keyboard files should be platform-independent, however, if included, a platform id should be the first subtag after `-t-k0-`.

Author: +1

| Platform | Source | Notes |
|----------|--------|-------|
| Android  | Android 4.0 - Ice Cream Sandwich ([https://source.android.com/source/downloading.html](https://source.android.com/source/downloading.html)) | Parsed layout files located in packages/inputmethods/LatinIME/java/res |
| Chrome OS | XKB ([https://www.x.org/wiki/XKB](https://www.x.org/wiki/XKB)) | The Chrome OS represents a very small subset of the keyboards available from XKB.

> | Mac OS X  | Ukelele bundled System Keyboards ([https://software.sil.org/ukelele/](https://software.sil.org/ukelele/)) | These layouts date from Mac OS X 10.4 and are therefore a bit outdated |

Reviewer 2: ```suggestion
| macOS     | Ukelele bundled System Keyboards ([https://software.sil.org/ukelele/](https://software.sil.org/ukelele/)) | These layouts date from Mac OS X 10.4 and are therefore a bit outdated |
```
